src/platform/i386/chal_pgtbl.c

```diff
@@ -0,0 +1,533 @@
+		/**
+                 * If no periodic flush done yet, did the
+		 * mandatory flush happen on all cores?
+                 */
+		for (i = 0; i < NUM_CPU_COS; i++) {
+			if (timestamp > tlb_quiescence[i].last_mandatory_flush) {
+				/* no go */
+				quiescent = 0;
+				break;
+			}
+		}
+	}
+	if (quiescent == 0) {
+		printk("from cpu %d, t %llu: cpu %d last mandatory flush: %llu\n", get_cpuid(), timestamp, i,
+		       tlb_quiescence[i].last_mandatory_flush);
+		for (i = 0; i < NUM_CPU_COS; i++) {
+			printk("cpu %d: flush %llu\n", i, tlb_quiescence[i].last_mandatory_flush);
+		}
+	}
+
+	return quiescent;
+}
+
+int
+chal_cap_memactivate(struct captbl *ct, struct cap_pgtbl *pt, capid_t frame_cap, capid_t dest_pt, vaddr_t vaddr)
+{
+	unsigned long *    pte, cosframe, orig_v;
+	struct cap_header *dest_pt_h;
+	u32_t              flags;
+	int                ret;
+
+	if (unlikely(pt->lvl || (pt->refcnt_flags & CAP_MEM_FROZEN_FLAG))) return -EINVAL;
+
+	dest_pt_h = captbl_lkup(ct, dest_pt);
+	if (dest_pt_h->type != CAP_PGTBL) return -EINVAL;
+	if (((struct cap_pgtbl *)dest_pt_h)->lvl) return -EINVAL;
+
+	pte = pgtbl_lkup_pte(pt->pgtbl, frame_cap, &flags);
+	if (!pte) return -EINVAL;
+	orig_v = *pte;
+
+	if (!(orig_v & X86_COSFRAME) || (orig_v & X86_COSKMEM)) return -EPERM;
+
+	assert(!(orig_v & X86_QUIESCENCE));
+	cosframe = orig_v & PGTBL_FRAME_MASK;
+
+	ret = pgtbl_mapping_add(((struct cap_pgtbl *)dest_pt_h)->pgtbl, vaddr, cosframe, X86_USER_DEF);
+
+	return ret;
+}
+
+int
+chal_pgtbl_activate(struct captbl *t, unsigned long cap, unsigned long capin, pgtbl_t pgtbl, u32_t lvl)
+{
+	struct cap_pgtbl *pt;
+	int               ret;
+
+	pt = (struct cap_pgtbl *)__cap_capactivate_pre(t, cap, capin, CAP_PGTBL, &ret);
+	if (unlikely(!pt)) return ret;
+	pt->pgtbl = pgtbl;
+
+	pt->refcnt_flags = 1;
+	pt->parent       = NULL; /* new cap has no parent. only copied cap has. */
+	pt->lvl          = lvl;
+	__cap_capactivate_post(&pt->h, CAP_PGTBL);
+
+	return 0;
+}
+
+int
+chal_pgtbl_deactivate(struct captbl *t, struct cap_captbl *dest_ct_cap, unsigned long capin,
+                      livenessid_t lid, capid_t pgtbl_cap, capid_t cosframe_addr, const int root)
+{
+	struct cap_header *deact_header;
+	struct cap_pgtbl * deact_cap, *parent;
+
+	unsigned long l, old_v = 0, *pte = NULL;
+	int           ret;
+
+	deact_header = captbl_lkup(dest_ct_cap->captbl, capin);
+	if (!deact_header || deact_header->type != CAP_PGTBL) cos_throw(err, -EINVAL);
+	deact_cap = (struct cap_pgtbl *)deact_header;
+	parent    = deact_cap->parent;
+
+	l = deact_cap->refcnt_flags;
+	assert(l & CAP_REFCNT_MAX);
+
+	if ((l & CAP_REFCNT_MAX) != 1) {
+		/* We need to deact children first! */
+		cos_throw(err, -EINVAL);
+	}
+
+	if (parent == NULL) {
+		if (!root) cos_throw(err, -EINVAL);
+		/**
+                 * Last reference to the captbl page. Require pgtbl
+		 * and cos_frame cap to release the kmem page.
+                 */
```

Comment:
why is the comment spacing off everywhere? i'll stop repeating.

Reply:
Drag your mouse while selecting the blank space.  You see it is a space vs. tab issue.

Reply:
Yes. Will change to all tabs then